## 4.9 Exercise Part 1
### This script will include
#### 00. Importing Libraries and Data
#### 01. Wrangle customer data
#### 02. Data Checks 
#### 03. Combine customer data with existing prepared data

### 00. Importing Libraries and Data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create folderpath shortcut
path = r'C:\Users\Victoria\Desktop\Data Analytics Career Foundry\4.0 Python Fundamentals for Data Analysts\Instacart Basket Analysis'

In [3]:
# Import customers.csv as dataframe customer

customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Import 'orders_products_aggregated.pkl' as ords_prods_merge

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

### 01. Wrangle customer data

Rename columns with illogical names and drop columns that don't add anything to your analysis

In [5]:
# Inspect new dataframe

customer.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [6]:
customer.shape

(206209, 10)

In [7]:
customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


#### Initial inspection

Upon first inspection, the 'First Name' column appears to be the only column with missing values.

There are many columns that need to be renamed. This could possibly result in a mixed-type column.

Descriptive statistics look normal and reasonable.

#### Renaming Columns

The following columns will be renamed:
1. First Name
2. Surnam
3. Gender
4. STATE
5. Age

I prefer all columns to be lowercase

In [9]:
# Renaming 'First Name' to 'first_name'

customer.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [10]:
# Renaming 'Surnam' to 'last_name'

customer.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [11]:
# Renaming 'Gender' to 'gender'

customer.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [12]:
# Renaming 'STATE' to 'state'

customer.rename(columns = {'STATE' : 'state'}, inplace = True)

In [13]:
# Renaming 'Age' to 'age'

customer.rename(columns = {'Age' : 'age'}, inplace = True)

In [14]:
# Confirming column renaming

customer.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

#### Dropping Columns

I will not be dropping any of the columns as I cannot determine that any of them wouldn't be useful at this time.

### 02. Data checks

Complete the fundamental data quality and consistency checks you’ve learned, check for and address missing values and duplicates, and convert any mixed-type data.

#### Checking for mixed-type data

In [15]:
# Checking data types

customer.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [16]:
#Check for mixed-types

for col in customer.columns.tolist():
    weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer[weird]) > 0:
        print(col)

first_name


In [17]:
# 'first_name' is a mixed-type because of the null values. I will change its data type to a string.

customer['first_name'] = customer['first_name'].astype('str')

In [18]:
# Confirming there are no mixed-types now

for col in customer.columns.tolist():
    weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer[weird]) > 0:
        print(col)

##### Conclusion

Mixed-types have been resolved

#### Checking missing values

In [19]:
# From the info() function I learned that 'first_name' has null values. I will investigate further now

customer['first_name'].value_counts(dropna = False)

nan        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: first_name, Length: 208, dtype: int64

In [20]:
# Confirming there are no missing values

customer.isnull().sum()

user_id         0
first_name      0
last_name       0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

##### Conclusion

There are 11,259 missing values within the 'first_name' column. So this column is missing (11,259/206209) or 5.46% of it's data. Without the original source data it would be impossible to fix these missing values. However, these values should not affect any statistics. I will leave them in for now. If the client has the original source data, it would be possible to fix them in the future.

### Checking for duplicates

In [21]:
# I will be looking for full duplicates

df_dups = customer[customer.duplicated()]

In [22]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

##### Conclusion

There are no duplicates.

### 03. Combine customer data with existing prepared data

In [23]:
# Checking ords_prods_merge

ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_the_week  int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  mean_price              float64 
 21  spendi

In [24]:
# Checking the ords_prods_merge dataframe size

ords_prods_merge.shape

(32404859, 24)

In [25]:
# Re-checking the customers.csv info

customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    206209 non-null  object
 2   last_name     206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [30]:
# Combining the two df using the 'user_id' column

opm_c_merge = ords_prods_merge.merge(customer, on = 'user_id')

In [31]:
# Check opm_c_merge

opm_c_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_the_week  int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                int64   
 11  department_id           int64   
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  mean_price              float64 
 21  spendi

In [32]:
#Checking shape to confirm number of rows is the same as ords_prods_merge
opm_c_merge.shape

(32404859, 33)

### 04. Export Data

In [33]:
# Export opm_c_merge as 'orders_products_customers_merged.pkl'

opm_c_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))